klippy/extras/firmware_retraction.py

```diff
@@ -0,0 +1,86 @@
+# Support for Marlin/Smoothie/Reprap style firmware retraction via G10/G11
+#
+# Copyright (C) 2019  Len Trigg <lenbok@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+class FirmwareRetraction:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.retract_length = config.getfloat('retract_length', 0., minval=0.)
+        self.retract_speed = config.getint('retract_speed', 2000, minval=1)
+        self.unretract_extra_length = config.getfloat(
+            'unretract_extra_length', 0., minval=0.)
+        self.unretract_speed = config.getint('unretract_speed', 500, minval=1)
+        self.z_hop = config.getfloat('z_hop', 0., minval=0.)
+        self.unretract_length = self.retract_length
+        + self.unretract_extra_length
+        self.is_retracted = False
+        self.gcode = self.printer.lookup_object('gcode')
+        self.gcode.register_command('SET_RETRACTION', self.cmd_SET_RETRACTION)
+        self.gcode.register_command('GET_RETRACTION', self.cmd_GET_RETRACTION)
+        self.gcode.register_command('G10', self.cmd_G10)
+        self.gcode.register_command('G11', self.cmd_G11)
+
+    def get_status(self, eventtime):
+        return {
+            "retract_length": self.retract_length,
+            "retract_speed": self.retract_speed,
+            "unretract_extra_length": self.unretract_extra_length,
+            "unretract_speed": self.unretract_speed,
+            "z_hop": self.z_hop
+        }
+
+    def cmd_SET_RETRACTION(self, params):
+        self.retract_length = self.gcode.get_float(
+            'RETRACT_LENGTH',
+            params, self.retract_length, minval=0.)
+        self.retract_speed = self.gcode.get_int(
```

Comment:
Use get_float() here.  Also, please use values in mm/s (not mm/min - g-code is just nuts).